src/pcie.c

```diff
@@ -0,0 +1,180 @@
+#include "utils.h"
+
+/*
+ * The ADT uses 17 register sets:
+ *
+ * 0:  90000000 00000006 10000000 00000000  ECAM
+ * 1:  80000000 00000006 00040000 00000000  RC
+ * 2:  80080000 00000006 00090000 00000000  PHY
+ * 3:  800c0000 00000006 00020000 00000000  PHY IP
+ * 4:  8c000000 00000006 00004000 00000000  AXI
+ * 5:  3d2bc000 00000000 00001000 00000000  fuses
+ * 6:  81000000 00000006 00008000 00000000  port 0
+ * 7:  81010000 00000006 00001000 00000000
+ * 8:  80084000 00000006 00004000 00000000  port 0 PHY
+ * 9:  800c8000 00000006 00016610 00000000
+ * 10: 82000000 00000006 00008000 00000000  port 1
+ * 11: 82010000 00000006 00001000 00000000
+ * 12: 80088000 00000006 00004000 00000000  port 1 PHY
+ * 13: 800d0000 00000006 00006000 00000000
+ * 14: 83000000 00000006 00008000 00000000  port 2
+ * 15: 83010000 00000006 00001000 00000000
+ * 16: 8008c000 00000006 00004000 00000000  port 2 PHY
+ * 17: 800d8000 00000006 00006000 00000000
+ */
+
+struct fuse_bits {
+    u16 src_reg;
+    u16 tgt_reg;
+    u8 src_bit;
+    u8 tgt_bit;
+    u8 width;
+};
+
+struct fuse_bits pcie_fuse_bits[] = {
+    {0x0084, 0x6238, 4, 0, 6},   {0x0084, 0x6220, 10, 14, 3}, {0x0084, 0x62a4, 13, 17, 2},
+    {0x0418, 0x522c, 27, 9, 2},  {0x0418, 0x522c, 13, 12, 3}, {0x0418, 0x5220, 18, 14, 3},
+    {0x0418, 0x52a4, 21, 17, 2}, {0x0418, 0x522c, 23, 16, 5}, {0x0418, 0x5278, 23, 20, 3},
+    {0x0418, 0x5018, 31, 2, 1},  {0x041c, 0x1204, 0, 2, 5},   {}};
+
+int pcie_init(void)
+{
+    int path[8];
+    int node = adt_path_offset_trace(adt, "/arm-io/apcie", path);
+
+    if (node < 0) {
+        printf("PCIE node not found!\n");
+        return -1;
+    }
+
+    u64 config_base;
+    if (adt_get_reg(adt, path, "reg", 0, &config_base, NULL)) {
+        printf("Failed to get PCIE reg property!\n");
+        return -1;
+    }
+
+    u64 rc_base;
+    if (adt_get_reg(adt, path, "reg", 1, &rc_base, NULL)) {
+        printf("Failed to get PCIE reg property!\n");
+        return -1;
+    }
+
+    u64 phy_base;
+    if (adt_get_reg(adt, path, "reg", 2, &phy_base, NULL)) {
+        printf("Failed to get PCIE reg property!\n");
+        return -1;
+    }
+
+    u64 phy_ip_base;
+    if (adt_get_reg(adt, path, "reg", 3, &phy_ip_base, NULL)) {
+        printf("Failed to get PCIE reg property!\n");
+        return -1;
+    }
+
+    u64 fuse_base;
+    if (adt_get_reg(adt, path, "reg", 5, &fuse_base, NULL)) {
+        printf("Failed to get PCIE reg property!\n");
+        return -1;
+    }
+
+    if (pmgr_adt_clocks_enable("/arm-io/apcie"))
+        return -1;
+
+    if (tunables_apply_local("/arm-io/apcie", "apcie-axi2af-tunables", 4))
+        return -1;
+    if (tunables_apply_local("/arm-io/apcie", "apcie-common-tunables", 1))
+        return -1;
+
+    /*
+     * Initialize PHY.
+     */
+
+    if (tunables_apply_local("/arm-io/apcie", "apcie-phy-tunables", 2))
+        return -1;
+
+    /* CLK0REQ/CLK0ACK */
+    set32(phy_base, 1);
+    if (poll32(phy_base, 4, 4, 50000))
```

Comment:
minor nit: if we know the register and bit names I'd `#define` them at the top

Reply:
I'm not 100% sure if all the names are correct and I'd have to come up with a meaningful name for the register too...
But yes, these need to be #defines.

Reply:
Added #defines for the register and bit names that we have some clues about.